### PR TITLE
For #3709: Add save to PDF UI.

### LINKED ISSUE #3709: [Spike] Save as PDF option
### Why/User Benefit/User Problem

To easily save a web page in PDF for formatting, portability and saving a local copy for later reading.

### What/Requirements

Save as PDF option to save the current webpage as a PDF file

### Acceptance Criteria (how do I know when I’m done?)

Feature parity with Fennec to save the current webpage as a PDF file

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-12323)

### PR DESCRIPTION
**Pull request by Amejia481**
https://www.figma.com/file/4TFvx6UlVO127bubv1WZKZ/Print-to-PDF-%2F-Share-Sheet?node-id=438%3A19971

image

We still need to land https://bugzilla.mozilla.org/show_bug.cgi?id=1784554 for improving the long time when the PDF is requested.

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #3709

### CODE REVIEW

**Comment by Amejia481** [comment]
Request for data collection review
---
1. What questions will you answer with this data?
    This data will provide statistics regarding the usage of the safe to PDF feature.
 
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
    This will give us an understanding of the usability of the button, info we will use for improving the user experience.
 
3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
    No alternative is available to answer this question.
 
4. Can current instrumentation answer these questions?
    No.
 
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement   Description | Data Collection   Category | Tracking Bug #
-- | -- | --
Home button clicks| category 2 - Interaction data | https://github.com/mozilla-mobile/fenix/issues/3709

6. How long will this data be collected? Choose one of the following:
    Data will be collected for 6 months initially. 
 
7. What populations will you measure?
    No filters - all locales, channels, etc. Currently, the button is visible only in Nightly.
 
8. If this data collection is default on, what is the opt-out mechanism for users?
     Default Glean SDK opt-out mechanism.
 
9. Please provide a general description of how you will analyze this data.
    Glean / Amplitude
 
10. Where do you intend to share the results of your analysis?
      Only on Glean, Amplitude and with mobile teams.
 
11. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
      No third-party tool will use this data.

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    onClick: () -> Unit,
+) {
+    Row(
+        modifier = Modifier
+            .height(48.dp)
+            .fillMaxWidth()
+            .clickable(onClick = onClick),
+    ) {
+        Column {
+            Divider(
```
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:40]: 🚗 nit: I'm wondering if it maybe makes more sense to to put this divider inline in the compose view in `ShareFragment`. This is just a thought for potential re-use: it seems like we may not want to assume a divider would be required anywhere this is used. It feels like the top-level UI would choose to use dividers to separate its child elements.

That said, I can see how that would be an awkward way to structure the code since its shared between XML and Compose. I don't have a strong opinion on what you choose to do, but thought it might be interesting food for thought.

File: app/src/test/java/org/mozilla/fenix/share/ShareControllerTest.kt [hunk 2]
```diff
@@ -133,8 +136,17 @@ class ShareControllerTest {
...
+            shareSubject,
+            shareData,
+            mockk(),
+            mockk(),
+            snackbar,
+            mockk(),
+            recentAppStorage,
+            testCoroutineScope,
+            testDispatcher,
+            dismiss,
```
Comment (MatthewTighe) [inline app/src/test/java/org/mozilla/fenix/share/ShareControllerTest.kt:149]: :car: nit: it would be helpful to update these with named parameters

File: app/src/test/java/org/mozilla/fenix/share/ShareInteractorTest.kt [hunk 3]
```diff
@@ -68,4 +68,11 @@ class ShareInteractorTest {
 
         verify { controller.handleShareToApp(app) }
     }
+
+    @Test
+    fun onSaveToPDF() {
+        interactor.onSaveToPDF("tabID")
+
+        verify { controller.handleSaveToPDF("tabID") }
+    }
```
Comment (MatthewTighe) [inline app/src/test/java/org/mozilla/fenix/share/ShareInteractorTest.kt:77]: :car: nit: I think we are trying to write all new tests in the `GIVEN WHEN THEN` style even if they do not match the rest of a file

File: app/src/main/res/layout/fragment_share.xml [hunk 4]
```diff
@@ -54,6 +54,14 @@
...
             app:layout_constraintBottom_toTopOf="@id/appsShareLayout" />
 
+        <androidx.compose.ui.platform.ComposeView
+            android:visibility="gone"
+            android:id="@+id/save_pdf"
+            android:layout_width="match_parent"
+            android:layout_height="72dp"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintTop_toBottomOf="@id/appsShareLayout" />
+
```
Comment (MatthewTighe) [inline app/src/main/res/layout/fragment_share.xml:64]: Or perhaps a divider in the XML that has conditional visibility like the ComposeView?

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                    ),
+                    tint = FirefoxTheme.colors.iconPrimary,
+                )
+
+                Text(
+                    stringResource(R.string.share_save_to_pdf),
+                )
+            }
+        }
+    }
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:60]: I noticed a few inaccuracies with the designs, so I ended updating them and adding them to this one block below:

```kotlin
    Row(
        modifier = Modifier
            .height(56.dp)
            .fillMaxWidth()
            .clickable(onClick = onClick),
    ) {
        Column {
            Divider(
                color = FirefoxTheme.colors.borderPrimary,
                thickness = 1.dp,
            )

            Row(
                modifier = Modifier.padding(top = 16.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(
                    modifier = Modifier.padding(start = 16.dp, end = 32.dp),
                    painter = painterResource(R.drawable.ic_download),
                    contentDescription = stringResource(
                        R.string.content_description_close_button,
                    ),
                    tint = FirefoxTheme.colors.iconPrimary,
                )

                Text(
                    stringResource(R.string.share_save_to_pdf),
                    color = FirefoxTheme.colors.textPrimary,
                )
            }
        }
    }
```

This is what it ends up looking like on my device:
Screen Shot 2022-10-03 at 11 08 41 PM

File: app/src/main/java/org/mozilla/fenix/share/ShareFragment.kt [hunk 6]
```diff
@@ -111,6 +116,19 @@ class ShareFragment : AppCompatDialogFragment() {
         }
         shareToAppsView = ShareToAppsView(binding.appsShareLayout, shareInteractor)
 
+        if (FeatureFlags.saveToPDF) {
+            binding.savePdf.isVisible = true
+            binding.savePdf.apply {
+                setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/share/ShareFragment.kt:122]: nit
```suggestion
            binding.savePdf.apply {
                isVisible = true
                setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
```

File: app/src/main/java/org/mozilla/fenix/share/ShareSaveToPDFInteractor.kt [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.share
+
+/**
+ * Callbacks for possible user interactions on the [SaveToPDFItem]
+ */
+interface ShareSaveToPDFInteractor {
+    /**
+     * Generates a PDF from given [tabId].
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/share/ShareSaveToPDFInteractor.kt:12]: typo
```suggestion
     * Generates a PDF from the given [tabId].
```

File: app/src/main/java/org/mozilla/fenix/share/ShareInteractor.kt [hunk 8]
```diff
@@ -12,7 +12,7 @@ import org.mozilla.fenix.share.listadapters.AppShareOption
  */
 class ShareInteractor(
     private val controller: ShareController,
-) : ShareCloseInteractor, ShareToAccountDevicesInteractor, ShareToAppsInteractor {
+) : ShareCloseInteractor, ShareToAccountDevicesInteractor, ShareToAppsInteractor, ShareSaveToPDFInteractor {
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/share/ShareInteractor.kt:15]: These class naming conventions do not make sense since we're already in the `share` package. 😄

File: app/src/main/res/layout/fragment_share.xml [hunk 9]
```diff
@@ -54,6 +54,14 @@
             android:background="?borderPrimary"
             app:layout_constraintBottom_toTopOf="@id/appsShareLayout" />
 
+        <androidx.compose.ui.platform.ComposeView
+            android:visibility="gone"
+            android:id="@+id/save_pdf"
+            android:layout_width="match_parent"
+            android:layout_height="72dp"
```
Comment (jonalmeida) [inline app/src/main/res/layout/fragment_share.xml:61]: This 72dp conflicts with the sizes in the composable dimensions (48dp in the code, but this is also wrong from the designs). It looks a bit off on my device with asymmetric padding (more on the bottom).

```suggestion
            android:layout_height="wrap_content"
```

**Review by jonalmeida — APPROVED** [review]
Mostly nits and few design fixes, but once they are fixed along with Matt's comments addressed, I think this looks good to land. Happy to re-review if needed. 🙂

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -1028,6 +1028,8 @@
     <!-- Content description (not visible, for screen readers etc.):
         "Share" button. Opens the share menu when pressed. -->
     <string name="share_button_content_description">Share</string>
+    <!-- Share to PDF button -->
```
Comment (jonalmeida) [inline app/src/main/res/values/strings.xml:1031]: nit: add some more detail description for translators to know where to find the feature.
```suggestion
    <!-- Text for the Save to PDF feature in the share sheet -->
```

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt (hunk 1 above)
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:40]: Similarly, if the `Divider` is re-located, the hierarchy could just be: 

```
Row(
   Icon()
   Text()
)

instead of a Row + Column + Row

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    ) {
+        Column {
+            Divider(
+                color = FirefoxTheme.colors.borderPrimary,
+                thickness = 1.dp,
+            )
+
+            Row(modifier = Modifier.padding(top = 16.dp)) {
+                Icon(
+                    modifier = Modifier.padding(start = 20.dp, end = 36.dp),
```
[the remark this answers is not in the document]
Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:47]: When items are clickable, IMO using padding instead of spacers would improve accessibility. (be aware of the modifiers order to keep padding clickable.)

[the remark this answers is not in the document]
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:47]: It's the entire row that's clickable here though, not the individual elements. The padding in this instance is just on the `Icon`.

[the remark this answers is not in the document]
Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:47]: Yeah, I was talking about the general "we lean towards" part, adding an instance where that would not be the best choice.

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+                    painter = painterResource(R.drawable.ic_download),
+                    contentDescription = stringResource(
+                        R.string.content_description_close_button,
+                    ),
+                    tint = FirefoxTheme.colors.iconPrimary,
+                )
+
+                Text(
+                    stringResource(R.string.share_save_to_pdf),
```
[the remark this answers is not in the document]
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:56]: Can you also add the named parameter and look for the appropriate `textStyle`

File: app/src/main/java/org/mozilla/fenix/share/ShareFragment.kt [hunk 13]
```diff
@@ -111,6 +116,19 @@ class ShareFragment : AppCompatDialogFragment() {
...
+                setContent {
+                    FirefoxTheme {
+                        SaveToPDFItem {
+                            shareInteractor.onSaveToPDF(args.sessionId)
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/share/ShareFragment.kt:126]: Can you add named parameters here

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt (hunk 1 above)
Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:40]: Yeah, I was wondering the same.  I'm not sure for this case if it's worth while trying to separated the divider from the UI, as there is not much to reuse 😅,  as we just have an icon and text, and it's customized to be part of the `ShareFragment` . I prefer to leave it as it's and wait until we can migrate the `ShareFragment` :)

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:40]: Sorry, @MozillaNoah I just refreshed and saw your message, thank for the suggestion!!

File: app/src/main/res/navigation/nav_graph.xml [hunk 14]
```diff
@@ -895,7 +895,7 @@
             app:destination="@id/addNewDeviceFragment" />
         <argument
             android:name="sessionId"
-            android:defaultValue="null"
+            android:defaultValue="@null"
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply app/src/main/res/navigation/nav_graph.xml:898]: And a painful one too, a lot of debugging 😆

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt (hunk 5 above)
Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt]: Thanks ❤️ !!

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt (hunk 1 above)
Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt]: Thanks both for the suggestions, I moved the divider to the xml as @MozillaNoah mention the compose UI would be much cleaner 😄.

**Comment by Amejia481** [comment]
Thanks all for all the feedback, I address the comment and updated the PR, when you have a moment please take a look :)

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+                )
+            }
+        }
+    }
+}
+
+@Composable
+@Preview
```
[the remark this answers is not in the document]
Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt:63]: I think you might have missed removing the first Preview annotation.

File: app/src/main/java/org/mozilla/fenix/share/ShareFragment.kt [hunk 16]
```diff
@@ -111,6 +116,20 @@ class ShareFragment : AppCompatDialogFragment() {
...
 
+        if (FeatureFlags.saveToPDF) {
+            binding.dividerLineAppsShareAndPdfSection.isVisible = true
+            binding.savePdf.apply {
+                isVisible = true
+                setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+                setContent {
+                    FirefoxTheme {
+                        SaveToPDFItem {
+                            shareInteractor.onSaveToPDF(tabId = args.sessionId)
```
Comment (mcarare) [inline app/src/main/java/org/mozilla/fenix/share/ShareFragment.kt:127]: nit: maybe it's just me, but it looks strange having the lambda out of the parenthesis, in a "tree" of composables content. 😅 . 
Maybe you could leave it on the same row as `SaveToPDFItem` so it looks more like a method call rather than an "inside" composable.

**Review by mcarare — APPROVED** [review]
LGTM!

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import org.mozilla.fenix.R
+import org.mozilla.fenix.theme.FirefoxTheme
+
+/**
+ * A save to PDF item.
+ * @param onClick event handler when the save to PDF item is clicked.
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt]: Let's fix the linter to enforce this if it's a valuable change we always want to have.

File: app/src/main/java/org/mozilla/fenix/share/Interactor.kt [hunk 18]
```diff
@@ -10,9 +10,9 @@ import org.mozilla.fenix.share.listadapters.AppShareOption
 /**
  * Interactor for the share screen.
  */
-class ShareInteractor(
+class Interactor(
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/share/Interactor.kt:13]: Was this change intentional?

File: app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/share/SaveToPDFItem.kt]: @jonalmeida I generally agree. If you have some knowledge about adding these changes to linters, can you ping me about it

File: app/src/main/java/org/mozilla/fenix/share/Interactor.kt (hunk 18 above)
Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/share/Interactor.kt:13]: Thanks for catching this it wasn't

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/share/Interactor.kt]: Addressed!
